cub/agent/agent_batch_memcpy.cuh

```diff
@@ -0,0 +1,1126 @@
+                                                         BlockBufferOffsetT num_wlev_buffers)
+  {
+    const int32_t warp_id              = threadIdx.x / CUB_PTX_WARP_THREADS;
+    const int32_t warp_lane            = threadIdx.x % CUB_PTX_WARP_THREADS;
+    constexpr uint32_t WARPS_PER_BLOCK = BLOCK_THREADS / CUB_PTX_WARP_THREADS;
+
+    for (BlockBufferOffsetT buffer_offset = warp_id; buffer_offset < num_wlev_buffers;
+         buffer_offset += WARPS_PER_BLOCK)
+    {
+      const auto buffer_id = buffers_by_size_class[buffer_offset].buffer_id;
+      detail::VectorizedCopy<CUB_PTX_WARP_THREADS, uint4>(warp_lane,
+                                                          tile_buffer_dsts[buffer_id],
+                                                          tile_buffer_sizes[buffer_id],
+                                                          tile_buffer_srcs[buffer_id]);
+    }
+  }
+
+  /**
+   * @brief Read in all the buffers of this tile that require thread-level collaboration and copy
+   * their bytes to the corresponding destination buffer
+   */
+  __device__ __forceinline__ void BatchMemcpyTLEVBuffers(BufferTuple *buffers_by_size_class,
+                                                         InputBufferIt tile_buffer_srcs,
+                                                         OutputBufferIt tile_buffer_dsts,
+                                                         BlockBufferOffsetT num_tlev_buffers)
+  {
+    // Read in the buffers' ids that require thread-level collaboration (where buffer id is the
+    // buffer within this tile)
+    BlockBufferOffsetT tlev_buffer_ids[TLEV_BUFFERS_PER_THREAD];
+    TLevBufferSizeT tlev_buffer_sizes[TLEV_BUFFERS_PER_THREAD];
+    // Currently we do not go over the TLEV buffers in multiple iterations, so we need to make sure
+    // we are able to be covered for the case that all our buffers are TLEV buffers
+    static_assert(TLEV_BUFFERS_PER_THREAD >= BUFFERS_PER_THREAD,
+                  "Unsupported confiugraiton: The number of 'thread-level buffers' must be at "
+                  "least as large as the number of overall buffers being processed by each "
+                  "thread.");
+
+    // Read in the TLEV buffer partition (i.e., the buffers that require thread-level collaboration)
+    uint32_t tlev_buffer_offset = threadIdx.x * TLEV_BUFFERS_PER_THREAD;
+
+    // Pre-populate the buffer sizes to 0 (i.e. zero-padding towards the end) to ensure
+    // out-of-bounds TLEV buffers will not be considered
+#pragma unroll
+    for (uint32_t i = 0; i < TLEV_BUFFERS_PER_THREAD; i++)
+    {
+      tlev_buffer_sizes[i] = 0;
+    }
+
+    // Assign TLEV buffers in a blocked arrangement (each thread is assigned consecutive TLEV
+    // buffers)
+#pragma unroll
+    for (uint32_t i = 0; i < TLEV_BUFFERS_PER_THREAD; i++)
+    {
+      if (tlev_buffer_offset < num_tlev_buffers)
+      {
+        tlev_buffer_ids[i]   = buffers_by_size_class[tlev_buffer_offset].buffer_id;
```

Comment:
I don't think there's a requirement to load this in blocked manner. Would stripped loads provide any advantage, maybe avoid bank conflicts?

Reply:
It's required for the `BlockRunLengthDecode` to function properly. From the docs:

> Trailing runs of length 0 are supported (i.e., they may only appear at the end of the run_lengths array). A run of length zero may not be followed by a run length that is not zero.

With a striped arrangement that couldn't be guaranteed.